### PR TITLE
Srsnv_motif_balancing_feature

### PR DESCRIPTION
**Pull request by kenissur**
Fixed motif balancing feature + minor fixes

### CODE REVIEW

File: ugvc/mrd/srsnv_training_utils.py [hunk 1]
```diff
@@ -165,10 +165,14 @@ def prepare_featuremap_for_model(
...
     if featuremap_entry_number < total_size:
         logger.warning(
-            f"featuremap_entry_number={featuremap_entry_number} > training_set_size={train_set_size}"
+            "Requested training and test set size cannot be met - insufficient data"
+            f"featuremap_entry_number={featuremap_entry_number} < training_set_size={train_set_size}"
             f"+ test_set_size={test_set_size if test_set_size else 0}"
-            "\nbehavior is undefined",
-        )  # TODO in this case we should make sure we still have a test set (and behavior should be defined)
+        )
+        if test_set_size is not None:
```
Comment (doron-st) [inline ugvc/mrd/srsnv_training_utils.py]: why would it be None?

File: ugvc/mrd/srsnv_training_utils.py [hunk 2]
```diff
@@ -165,10 +165,14 @@ def prepare_featuremap_for_model(
...
+            train_set_size = np.floor(featuremap_entry_number * (train_set_size / total_size))
```
Comment (doron-st) [inline ugvc/mrd/srsnv_training_utils.py:173]: the total_size is the total size of the featuremap?
So if test-size is None we use all remaining rows as test?

File: test/unit/mrd/test_srsnv_train_utils.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def test_prepare_featuremap_for_model_with_motif_balancing(tmpdir):
+    """Test that downsampling training-set works as expected"""
+
+    input_featuremap_vcf = pjoin(
+        inputs_dir,
+        "333_LuNgs_08.annotated_featuremap.vcf.gz",
+    )
+    balanced_sampling_info_fields = ["trinuc_context", "is_forward"]
```
Comment (doron-st) [inline test/unit/mrd/test_srsnv_train_utils.py:50]: do we still have trinuc_contextas a feature? I'm confused.

File: test/unit/mrd/test_srsnv_train_utils.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        )
+        assert __count_variants(downsampled_training_featuremap_vcf) == train_set_size
+
+        balanced_sampling_info_fields_counter = defaultdict(int)
+        with pysam.VariantFile(downsampled_training_featuremap_vcf) as fmap:
+            for record in fmap.fetch():
+                balanced_sampling_info_fields_counter[
+                    tuple(record.info.get(info_field) for info_field in balanced_sampling_info_fields)
+                ] += 1
+        assert sum(balanced_sampling_info_fields_counter.values()) == train_set_size
```
Comment (doron-st) [inline test/unit/mrd/test_srsnv_train_utils.py:69]: Let's add an assertion about the uniformity

File: ugvc/pipelines/mrd/srsnv_training.py [hunk 5]
```diff
@@ -83,6 +83,12 @@ def parse_args(argv: list[str]) -> argparse.Namespace:
         type=str,
         help="""comma separated list of categorical features for ML classifier """,
     )
+    parser.add_argument(
+        "--balanced_sampling_info_fields",
+        type=str,
+        help="comma separated list of categorical features to be used for balanced sampling of the TP training set"
+        " to eliminate prior distribution bias (e.g. 'trinuc_context_with_alt,is_forward')",
```
Comment (doron-st) [inline ugvc/pipelines/mrd/srsnv_training.py:90]: remind me, the context of forward and reverse reads on the same snv is the same or revcom?

**Review by doron-st — COMMENTED** [review]
Looks mostly good. A few questions and a small request on the test

**Comment by doron-st** [comment]
A general question about balancing.
Do you balance only the TPs or all?

File: test/unit/mrd/test_srsnv_train_utils.py (hunk 3 above)
Comment (kenissur) [reply test/unit/mrd/test_srsnv_train_utils.py:50]: Oh. There's an older file... do you insist I change it?

File: ugvc/mrd/srsnv_training_utils.py (hunk 1 above)
Comment (kenissur) [reply ugvc/mrd/srsnv_training_utils.py]: No reason, fixing

File: ugvc/mrd/srsnv_training_utils.py (hunk 2 above)
Comment (kenissur) [inline ugvc/mrd/srsnv_training_utils.py]: total_size = train_set_size + test_set_size

Removing the option to have None

File: ugvc/pipelines/mrd/srsnv_training.py (hunk 5 above)
Comment (kenissur) [reply ugvc/pipelines/mrd/srsnv_training.py:90]: The same, that's why I added is_forward

File: test/unit/mrd/test_srsnv_train_utils.py (hunk 4 above)
Comment (kenissur) [reply test/unit/mrd/test_srsnv_train_utils.py:69]: How is that? :)

File: ugvc/mrd/srsnv_training_utils.py (hunk 2 above)
Comment (doron-st) [reply ugvc/mrd/srsnv_training_utils.py]: Wait. 
if featuremap_entry_number is the size of the featuremap (in the region)
and total_size is train+test 
this calculation seems weird isn't it?

Comment (kenissur) [reply ugvc/mrd/srsnv_training_utils.py]: I think it's okay. I am handling the case where we asked for more training data than the total number of available entries, so in this case I am using approximately all the data, but maintaining the ratio of train_set_size to test_set_size

File: ugvc/mrd/srsnv_training_utils.py [hunk 6]
```diff
@@ -185,28 +189,23 @@ def prepare_featuremap_for_model(
...
         np.random.seed(random_seed)
     if balanced_sampling_info_fields:
-        # TODO test this code and add a test, I wrote an implementation but didn't even get to run it
-        logger.warning(
-            "Data motif balancing not tested, use at your own risk"
-            f" (balanced_sampling_info_fields={balanced_sampling_info_fields})"
-        )
         # count the number of entries with each combination of info fields to balance by
         balanced_sampling_info_fields_counter = defaultdict(int)
         with pysam.VariantFile(intersect_featuremap_vcf) as fmap:
```
Comment (doron-st) [inline ugvc/mrd/srsnv_training_utils.py]: Wouldn't it be better to count while in the previous general counting loop? for efficiency reasons.

File: ugvc/mrd/srsnv_training_utils.py [hunk 7]
```diff
@@ -185,28 +189,23 @@ def prepare_featuremap_for_model(
...
             for record in fmap.fetch():
                 balanced_sampling_info_fields_counter[
-                    (record.info.get(info_field) for info_field in balanced_sampling_info_fields_counter)
+                    tuple(record.info.get(info_field) for info_field in balanced_sampling_info_fields)
                 ] += 1
         # determine sampling rate per group (defined by a combination of info fields)
         total_size_per_group = total_size // len(balanced_sampling_info_fields_counter)
```
Comment (doron-st) [inline ugvc/mrd/srsnv_training_utils.py]: I'm confused by this calculation.
Isn't it more natural to count the fraction of events with this info-fields from the total?
And then use the inverse of that to drive the balancing?

File: ugvc/mrd/srsnv_training_utils.py [hunk 8]
```diff
@@ -185,28 +189,23 @@ def prepare_featuremap_for_model(
...
-            k: total_size_per_group / v * overhead_factor for k, v in balanced_sampling_info_fields_counter.items()
+            k: (total_size_per_group / v) * overhead_factor for k, v in balanced_sampling_info_fields_counter.items()
         }
         for k, v in downsampling_rate.items():
             logger.debug(f"downsampling_rate for {k} = {v}")
-            if v > 1:
+            if v > overhead_factor:
                 logger.warning(f"downsampling_rate for {k} = {v} > 1, result will contain less data than expected")
-                downsampling_rate[k] = 1
+                downsampling_rate[k] = min(1, downsampling_rate[k])
```
Comment (doron-st) [inline ugvc/mrd/srsnv_training_utils.py]: Still not sure what's going on here. sorry...
If we want to make sure ds rate is not larger than 1, why don't we do this explicitly?
downsampling_rate[k] = min(1, downsampling_rate[k])

File: ugvc/mrd/srsnv_training_utils.py (hunk 6 above)
Comment (kenissur) [reply ugvc/mrd/srsnv_training_utils.py]: You're right, fixed.

File: ugvc/mrd/srsnv_training_utils.py (hunk 7 above)
Comment (kenissur) [reply ugvc/mrd/srsnv_training_utils.py]: 1. I removed the overhead factor for simplicity
2. total_size_per_group is the total number of the entries we want in the output, and I am dividing it by the total number of groups we are balancing across to get total_size_per_group
3. For each group I am then dividing total_size_per_group by the number of total number of featuremap entries in the group to get the rate at which I want to downsample. I made the variable names more explicit.

File: ugvc/mrd/srsnv_training_utils.py (hunk 8 above)
Comment (kenissur) [reply ugvc/mrd/srsnv_training_utils.py]: It is done explicitly in the next line

File: ugvc/pipelines/mrd/srsnv_training.py (hunk 5 above)
Comment (doron-st) [reply ugvc/pipelines/mrd/srsnv_training.py:90]: So your intent is that in the end we will have the same number of +/- reads per motif?
I'm not sure why wouldn't it occur naturally in germline variants.

**Review by doron-st — APPROVED** [review]
Excellent work

**Pull request merged**